// xdl/build/Analytics.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticsClient = void 0;
const analytics_node_1 = __importDefault(require("analytics-node"));
const os_1 = __importDefault(require("os"));
const internal_1 = require("./internal");
const PLATFORM_TO_ANALYTICS_PLATFORM = {
    darwin: 'Mac',
    win32: 'Windows',
    linux: 'Linux',
};
class AnalyticsClient {
    get userId() {
        return this._userId;
    }
    get version() {
        return this._version;
    }
    flush() {
        if (this.segmentNodeInstance) {
            this.segmentNodeInstance.flush();
        }
    }
    initializeClient(apiKey, packageVersion) {
        // Do not wait before flushing, we want node to close immediately if the programs ends
        this.segmentNodeInstance = new analytics_node_1.default(apiKey, { flushInterval: 300 });
        this._version = packageVersion;
    }
    identifyUser(userId, traits) {
        this._userId = userId;
        this.userTraits = traits;
        if (this.segmentNodeInstance) {
            this.segmentNodeInstance.identify({
                userId: this._userId,
                traits: this.userTraits,
                context: this.getContext(),
            });
        }
    }
    logEvent(name, properties = {}) {
        if (this.segmentNodeInstance && this._userId) {
            this.segmentNodeInstance.track({
                userId: this._userId,
                event: name,
                properties,
                context: this.getContext(),
            });
        }
    }
    getContext() {
        const platform = PLATFORM_TO_ANALYTICS_PLATFORM[os_1.default.platform()] || os_1.default.platform();
        const context = {
            ip: internal_1.ip.address(),
            device: {
                model: platform,
                brand: platform,
            },
            os: {
                name: platform,
                version: os_1.default.release(),
            },
            app: {},
        };
        if (this._version) {
            context.app = {
                version: this._version,
            };
        }
        return context;
    }
}
exports.AnalyticsClient = AnalyticsClient;
const defaultClient = new AnalyticsClient();
exports.default = defaultClient;
