// xdl/build/internal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Extract = exports.ImageUtils = exports.ip = exports.IosCodeSigning = exports.Detach = exports.ModuleVersion = exports.ProjectUtils = exports.Doctor = exports.PackagerLogsStream = exports.PKCS12Utils = exports.AndroidCredentials = exports.XDLError = exports.Webpack = exports.Versions = exports.UserSettings = exports.ANONYMOUS_USERNAME = exports.UserManagerInstance = exports.UserManager = exports.UrlUtils = exports.ThirdParty = exports.StandaloneBuild = exports.Simulator = exports.SimControlLogs = exports.SimControl = exports.Prompts = exports.ProjectSettings = exports.ProjectAssets = exports.writeArtifactSafelyAsync = exports.prepareHooks = exports.runHook = exports.getPublishExpConfigAsync = exports.printBundleSizes = exports.createBundlesAsync = exports.publishAsync = exports.Exp = exports.EmbeddedAssets = exports.Binaries = exports.ApiV2 = exports.Android = exports.UnifiedAnalytics = exports.AnalyticsClient = exports.Analytics = exports.learnMore = exports.NotificationCode = exports.Logger = exports.ConnectionStatus = exports.Xcode = exports.Config = exports.Env = exports.Semaphore = void 0;
exports.Project = exports.getFreePortAsync = exports.ManifestHandler = exports.broadcastMessage = exports.stopAsync = exports.startAsync = exports.startDevServerAsync = exports.resolveNgrokAsync = exports.DevSession = exports.Sentry = exports.Watchman = exports.resolveEntryPoint = exports.TableText = exports.AssetBundle = exports.StandaloneBuildFlags = exports.pipeOutputToLogger = exports.LoggerDetach = exports.WebpackCompiler = exports.WebpackEnvironment = exports.FsCache = exports.BundleIdentifier = exports.downloadAppAsync = exports.downloadApkAsync = exports.choosePortAsync = exports.delayAsync = exports.ExpSchema = exports.stopReactNativeServerAsync = exports.startReactNativeServerAsync = exports.stopExpoServerAsync = exports.startExpoServerAsync = exports.stopTunnelsAsync = exports.startTunnelsAsync = exports.assertValidProjectRoot = exports.IosWorkspace = exports.IosPlist = exports.ExponentTools = exports.StandaloneContext = exports.Session = void 0;
/*
  All imports of XDL modules *within* XDL should be imported from this module.
  This file exists so that we can control the loading order of modules: without
  it circular dependencies that are used during module execution will cause
  hard-to-fix errors where the constant or function imported is undefined,
  because the module involved in a circular dependency hasn't been fully loaded
  at the time it's used.

  With this module in place, we can fix these circular dependency problems by
  changing the order in this file.
*/
var Semaphore_1 = require("./utils/Semaphore");
Object.defineProperty(exports, "Semaphore", { enumerable: true, get: function () { return Semaphore_1.Semaphore; } });
exports.Env = __importStar(require("./Env"));
var Config_1 = require("./Config");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return __importDefault(Config_1).default; } });
exports.Xcode = __importStar(require("./Xcode"));
exports.ConnectionStatus = __importStar(require("./ConnectionStatus"));
var Logger_1 = require("./Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return __importDefault(Logger_1).default; } });
var NotificationCode_1 = require("./NotificationCode");
Object.defineProperty(exports, "NotificationCode", { enumerable: true, get: function () { return __importDefault(NotificationCode_1).default; } });
var TerminalLink_1 = require("./logs/TerminalLink");
Object.defineProperty(exports, "learnMore", { enumerable: true, get: function () { return TerminalLink_1.learnMore; } });
var Analytics_1 = require("./Analytics");
Object.defineProperty(exports, "Analytics", { enumerable: true, get: function () { return __importDefault(Analytics_1).default; } });
Object.defineProperty(exports, "AnalyticsClient", { enumerable: true, get: function () { return Analytics_1.AnalyticsClient; } });
var UnifiedAnalytics_1 = require("./UnifiedAnalytics");
Object.defineProperty(exports, "UnifiedAnalytics", { enumerable: true, get: function () { return __importDefault(UnifiedAnalytics_1).default; } });
exports.Android = __importStar(require("./Android"));
var ApiV2_1 = require("./ApiV2");
Object.defineProperty(exports, "ApiV2", { enumerable: true, get: function () { return __importDefault(ApiV2_1).default; } });
exports.Binaries = __importStar(require("./Binaries"));
exports.EmbeddedAssets = __importStar(require("./EmbeddedAssets"));
exports.Exp = __importStar(require("./Exp"));
var publishAsync_1 = require("./project/publishAsync");
Object.defineProperty(exports, "publishAsync", { enumerable: true, get: function () { return publishAsync_1.publishAsync; } });
var createBundlesAsync_1 = require("./project/createBundlesAsync");
Object.defineProperty(exports, "createBundlesAsync", { enumerable: true, get: function () { return createBundlesAsync_1.createBundlesAsync; } });
Object.defineProperty(exports, "printBundleSizes", { enumerable: true, get: function () { return createBundlesAsync_1.printBundleSizes; } });
var getPublishExpConfigAsync_1 = require("./project/getPublishExpConfigAsync");
Object.defineProperty(exports, "getPublishExpConfigAsync", { enumerable: true, get: function () { return getPublishExpConfigAsync_1.getPublishExpConfigAsync; } });
var runHook_1 = require("./project/runHook");
Object.defineProperty(exports, "runHook", { enumerable: true, get: function () { return runHook_1.runHook; } });
Object.defineProperty(exports, "prepareHooks", { enumerable: true, get: function () { return runHook_1.prepareHooks; } });
var ArtifactUtils_1 = require("./tools/ArtifactUtils");
Object.defineProperty(exports, "writeArtifactSafelyAsync", { enumerable: true, get: function () { return ArtifactUtils_1.writeArtifactSafelyAsync; } });
exports.ProjectAssets = __importStar(require("./ProjectAssets"));
exports.ProjectSettings = __importStar(require("./ProjectSettings"));
exports.Prompts = __importStar(require("./Prompts"));
exports.SimControl = __importStar(require("./SimControl"));
exports.SimControlLogs = __importStar(require("./SimControlLogs"));
exports.Simulator = __importStar(require("./Simulator"));
exports.StandaloneBuild = __importStar(require("./StandaloneBuild"));
exports.ThirdParty = __importStar(require("./ThirdParty"));
exports.UrlUtils = __importStar(require("./UrlUtils"));
var User_1 = require("./User");
Object.defineProperty(exports, "UserManager", { enumerable: true, get: function () { return __importDefault(User_1).default; } });
Object.defineProperty(exports, "UserManagerInstance", { enumerable: true, get: function () { return User_1.UserManagerInstance; } });
Object.defineProperty(exports, "ANONYMOUS_USERNAME", { enumerable: true, get: function () { return User_1.ANONYMOUS_USERNAME; } });
var UserSettings_1 = require("./UserSettings");
Object.defineProperty(exports, "UserSettings", { enumerable: true, get: function () { return __importDefault(UserSettings_1).default; } });
exports.Versions = __importStar(require("./Versions"));
exports.Webpack = __importStar(require("./Webpack"));
var XDLError_1 = require("./XDLError");
Object.defineProperty(exports, "XDLError", { enumerable: true, get: function () { return __importDefault(XDLError_1).default; } });
exports.AndroidCredentials = __importStar(require("./credentials/AndroidCredentials"));
exports.PKCS12Utils = __importStar(require("./detach/PKCS12Utils"));
var PackagerLogsStream_1 = require("./logs/PackagerLogsStream");
Object.defineProperty(exports, "PackagerLogsStream", { enumerable: true, get: function () { return __importDefault(PackagerLogsStream_1).default; } });
exports.Doctor = __importStar(require("./project/Doctor"));
exports.ProjectUtils = __importStar(require("./project/ProjectUtils"));
exports.ModuleVersion = __importStar(require("./tools/ModuleVersion"));
exports.Detach = __importStar(require("./detach/Detach"));
exports.IosCodeSigning = __importStar(require("./detach/IosCodeSigning"));
var ip_1 = require("./ip");
Object.defineProperty(exports, "ip", { enumerable: true, get: function () { return __importDefault(ip_1).default; } });
exports.ImageUtils = __importStar(require("./tools/ImageUtils"));
exports.Extract = __importStar(require("./Extract"));
exports.Session = __importStar(require("./Session"));
var StandaloneContext_1 = require("./detach/StandaloneContext");
Object.defineProperty(exports, "StandaloneContext", { enumerable: true, get: function () { return __importDefault(StandaloneContext_1).default; } });
exports.ExponentTools = __importStar(require("./detach/ExponentTools"));
exports.IosPlist = __importStar(require("./detach/IosPlist"));
exports.IosWorkspace = __importStar(require("./detach/IosWorkspace"));
var errors_1 = require("./project/errors");
Object.defineProperty(exports, "assertValidProjectRoot", { enumerable: true, get: function () { return errors_1.assertValidProjectRoot; } });
var ngrok_1 = require("./start/ngrok");
Object.defineProperty(exports, "startTunnelsAsync", { enumerable: true, get: function () { return ngrok_1.startTunnelsAsync; } });
Object.defineProperty(exports, "stopTunnelsAsync", { enumerable: true, get: function () { return ngrok_1.stopTunnelsAsync; } });
var startLegacyExpoServerAsync_1 = require("./start/startLegacyExpoServerAsync");
Object.defineProperty(exports, "startExpoServerAsync", { enumerable: true, get: function () { return startLegacyExpoServerAsync_1.startExpoServerAsync; } });
Object.defineProperty(exports, "stopExpoServerAsync", { enumerable: true, get: function () { return startLegacyExpoServerAsync_1.stopExpoServerAsync; } });
var startLegacyReactNativeServerAsync_1 = require("./start/startLegacyReactNativeServerAsync");
Object.defineProperty(exports, "startReactNativeServerAsync", { enumerable: true, get: function () { return startLegacyReactNativeServerAsync_1.startReactNativeServerAsync; } });
Object.defineProperty(exports, "stopReactNativeServerAsync", { enumerable: true, get: function () { return startLegacyReactNativeServerAsync_1.stopReactNativeServerAsync; } });
exports.ExpSchema = __importStar(require("./project/ExpSchema"));
var delayAsync_1 = require("./utils/delayAsync");
Object.defineProperty(exports, "delayAsync", { enumerable: true, get: function () { return delayAsync_1.delayAsync; } });
var choosePortAsync_1 = require("./utils/choosePortAsync");
Object.defineProperty(exports, "choosePortAsync", { enumerable: true, get: function () { return choosePortAsync_1.choosePortAsync; } });
var downloadApkAsync_1 = require("./utils/downloadApkAsync");
Object.defineProperty(exports, "downloadApkAsync", { enumerable: true, get: function () { return downloadApkAsync_1.downloadApkAsync; } });
var downloadAppAsync_1 = require("./utils/downloadAppAsync");
Object.defineProperty(exports, "downloadAppAsync", { enumerable: true, get: function () { return downloadAppAsync_1.downloadAppAsync; } });
exports.BundleIdentifier = __importStar(require("./BundleIdentifier"));
exports.FsCache = __importStar(require("./tools/FsCache"));
exports.WebpackEnvironment = __importStar(require("./webpack-utils/WebpackEnvironment"));
exports.WebpackCompiler = __importStar(require("./webpack-utils/WebpackCompiler"));
var Logger_2 = require("./detach/Logger");
Object.defineProperty(exports, "LoggerDetach", { enumerable: true, get: function () { return __importDefault(Logger_2).default; } });
Object.defineProperty(exports, "pipeOutputToLogger", { enumerable: true, get: function () { return Logger_2.pipeOutputToLogger; } });
var StandaloneBuildFlags_1 = require("./detach/StandaloneBuildFlags");
Object.defineProperty(exports, "StandaloneBuildFlags", { enumerable: true, get: function () { return __importDefault(StandaloneBuildFlags_1).default; } });
exports.AssetBundle = __importStar(require("./detach/AssetBundle"));
exports.TableText = __importStar(require("./logs/TableText"));
var resolveEntryPoint_1 = require("./tools/resolveEntryPoint");
Object.defineProperty(exports, "resolveEntryPoint", { enumerable: true, get: function () { return resolveEntryPoint_1.resolveEntryPoint; } });
exports.Watchman = __importStar(require("./Watchman"));
exports.Sentry = __importStar(require("./Sentry"));
exports.DevSession = __importStar(require("./DevSession"));
var resolveNgrok_1 = require("./start/resolveNgrok");
Object.defineProperty(exports, "resolveNgrokAsync", { enumerable: true, get: function () { return resolveNgrok_1.resolveNgrokAsync; } });
var startDevServerAsync_1 = require("./start/startDevServerAsync");
Object.defineProperty(exports, "startDevServerAsync", { enumerable: true, get: function () { return startDevServerAsync_1.startDevServerAsync; } });
var startAsync_1 = require("./start/startAsync");
Object.defineProperty(exports, "startAsync", { enumerable: true, get: function () { return startAsync_1.startAsync; } });
Object.defineProperty(exports, "stopAsync", { enumerable: true, get: function () { return startAsync_1.stopAsync; } });
Object.defineProperty(exports, "broadcastMessage", { enumerable: true, get: function () { return startAsync_1.broadcastMessage; } });
exports.ManifestHandler = __importStar(require("./start/ManifestHandler"));
var getFreePortAsync_1 = require("./start/getFreePortAsync");
Object.defineProperty(exports, "getFreePortAsync", { enumerable: true, get: function () { return getFreePortAsync_1.getFreePortAsync; } });
exports.Project = __importStar(require("./Project"));
