// xdl/build/UserSettings.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const json_file_1 = __importDefault(require("@expo/json-file"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const uuid_1 = __importDefault(require("uuid"));
const internal_1 = require("./internal");
const SETTINGS_FILE_NAME = 'state.json';
function userSettingsFile() {
    return path_1.default.join(dotExpoHomeDirectory(), SETTINGS_FILE_NAME);
}
function userSettingsJsonFile() {
    return new json_file_1.default(userSettingsFile(), {
        jsonParseErrorDefault: {},
        cantReadFileDefault: {},
    });
}
let mkdirped = false;
function dotExpoHomeDirectory() {
    let dirPath = process.env.__UNSAFE_EXPO_HOME_DIRECTORY;
    if (!dirPath) {
        const home = os_1.default.homedir();
        if (internal_1.Env.isStaging()) {
            dirPath = path_1.default.join(home, '.expo-staging');
        }
        else if (internal_1.Env.isLocal()) {
            dirPath = path_1.default.join(home, '.expo-local');
        }
        else {
            dirPath = path_1.default.join(home, '.expo');
        }
    }
    if (!mkdirped) {
        fs_extra_1.default.mkdirpSync(dirPath);
        mkdirped = true;
    }
    return dirPath;
}
// returns an anonymous, unique identifier for a user on the current computer
async function getAnonymousIdentifierAsync() {
    const settings = await userSettingsJsonFile();
    let id = await settings.getAsync('uuid', null);
    if (!id) {
        id = uuid_1.default.v4();
        await settings.setAsync('uuid', id);
    }
    return id;
}
function accessToken() {
    return process.env.EXPO_TOKEN || null;
}
const UserSettings = Object.assign(userSettingsJsonFile(), {
    dotExpoHomeDirectory,
    userSettingsFile,
    userSettingsJsonFile,
    accessToken,
    getAnonymousIdentifierAsync,
    SETTINGS_FILE_NAME,
});
exports.default = UserSettings;
